### DISCUSSION TITLE
How to rank autocomple result based on inline Completion result?

### DISCUSSION THREAD

**Discussion by yanmofeixi** [microsoft/vscode-discussions#1264] [Extension Development QnA] [answered]
Hello, I have two language servers, one registers autocomplete provider, the other registers inline completion provider, how do I rank the autocomplete's completion items based on inline completion result?

Currently it uses the first item in completion item list to trigger another inline completion, I believe it's because VS Code want the ghost text and completion item to have same prefix. cc @hediet

**Comment by theReynald** [comment, marked as answer]
Hi @yanmofeixi,

According to copilot, to rank autocomplete's completion items based on inline completion results in Visual Studio Code, where one language server registers an autocomplete provider and another registers an inline completion provider, you'll need to coordinate between the two servers since VS Code's API doesn't directly support cross-communication for ranking purposes. Here's a high-level approach:

1. Inline Completion Provider: Have the inline completion provider generate its completions with unique identifiers or markers that can be recognized later. This provider should prioritize its completions based on your internal logic.

2. Autocomplete Provider: In the autocomplete provider, when generating completion items, include a step to query or wait for inline completion results (this might involve a shared state or direct communication between your servers, depending on your architecture). Use the information from the inline completions to rank your autocomplete suggestions. This could involve matching the suggestions from the inline completion provider with those from the autocomplete provider and adjusting the sortText property of each CompletionItem to control the order they appear in.
 
3. Shared State or Communication Channel: Establish a shared state or a communication channel between the two servers. This could be a simple in-memory structure (if both servers run in the same process) or a more complex IPC mechanism (if they're separate). The key is to allow the autocomplete provider to access results from the inline completion provider in real-time or near-real-time.
 
4. Synchronization and Timing: Ensure that the timing of the autocomplete and inline completion providers is synchronized in such a way that the autocomplete provider can reliably access the latest inline completion results when it needs to generate its completion list.
 
5. Adjusting sortText: Use the sortText property of the CompletionItem objects in the autocomplete provider to rank the items based on the inline completion results. Items with a sortText value that appears earlier in the sort order will be ranked higher in the list.

Here's a simplified example of adjusting sortText based on some external criteria (in your case, the results from the inline completion provider):

```TypeScript
// Autocomplete provider
vscode.languages.registerCompletionItemProvider('yourLanguage', {
  provideCompletionItems(document, position) {
    const completionItems = []; // Assume this is filled with your items
    // Example: Adjusting sortText based on some external criteria
    completionItems.forEach(item => {
      // Your logic here to determine rank based on inline completion results
      // For demonstration, let's assume higher rank means earlier in the alphabet
      const rank = getRankFromInlineCompletion(item); // Implement this
      item.sortText = rank.toString().padStart(5, '0'); // Ensure a consistent sort order
    });
    return completionItems;
  }
});
```

Let me know if this helps & reopen if you have more questions.
